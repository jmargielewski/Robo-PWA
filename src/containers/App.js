import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setSearchField, requestRobots } from '../actions';

import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundry from '../components/ErrorBoundry';
import Header from '../components/Header';

import './App.css';

export class App extends Component {
  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = () => 
    this.props.robots.filter(robot => {
      return robot.name.toLowerCase().includes(this.props.searchField.toLowerCase());
    });

  render() {
    const { onSearchChange, isPending } = this.props;
    
    return (
      <div className='tc'>
        <Header />
        <SearchBox searchChange={onSearchChange}/>
        <Scroll>
          { isPending ? <h1>Loading</h1> :
            <ErrorBoundry>
              <CardList robots={this.filterRobots()} />
            </ErrorBoundry>
          }
        </Scroll>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    searchField: state.searchRobots.searchField,
    robots: state.requestRobots.robots,
    isPending: state.requestRobots.isPending
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onSearchChange: (event) => dispatch(setSearchField(event.target.value)),
    onRequestRobots: () => dispatch(requestRobots())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
