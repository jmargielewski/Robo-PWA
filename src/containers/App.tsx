import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setSearchField, requestRobots } from '../actions';

import CardList from '../components/CardList';
import SearchBox from '../components/SearchBox';
import Scroll from '../components/Scroll';
import ErrorBoundry from '../components/ErrorBoundry';
import Header from '../components/Header';

import './App.css';

export interface IRobot {
  name: string;
  id: number;
  email: string;
}

interface IAppProps {
  onRequestRobots: () => {};
  onSearchChange(event: React.SyntheticEvent<HTMLInputElement>): void;
  searchField: string;
  isPending: boolean;
  robots: Array<IRobot>,
}

export class App extends Component<IAppProps> {
  componentDidMount(): void {
    this.props.onRequestRobots();
  }

  filterRobots = (): IRobot[] => 
    this.props.robots.filter((robot: IRobot) =>
      robot.name.toLowerCase().includes(this.props.searchField.toLowerCase()));

  render(): JSX.Element {
    const { onSearchChange, isPending } = this.props;
    
    return (
      <div className='tc'>
        <Header />
        <SearchBox searchChange={onSearchChange}/>
        <Scroll>
          { isPending ? <h1>Loading</h1> :
            <ErrorBoundry>
              <CardList robots={this.filterRobots()} />
            </ErrorBoundry>
          }
        </Scroll>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    searchField: state.searchRobots.searchField,
    robots: state.requestRobots.robots,
    isPending: state.requestRobots.isPending
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    onSearchChange: (event) => dispatch(setSearchField(event.target.value)),
    onRequestRobots: () => dispatch(requestRobots())
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
