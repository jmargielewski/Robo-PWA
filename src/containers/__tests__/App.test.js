import React from 'react';
import { shallow } from 'enzyme';
import { App } from '../App';
import toJson from 'enzyme-to-json';
// npm redux-mock-store

let wrapper;

beforeEach(() => {
    const mockProps = {
        onRequestRobots: jest.fn(),
        robots: [],
        searchField: '',
        isPending: false,
    };
    wrapper = shallow(<App { ...mockProps } />);
});

it('render App without crashing', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
});

it('filters robots correctly', () => {
    const mockProps2 = {
        onRequestRobots: jest.fn(),
        robots: [{
            id: 1,
            name: 'Robo 1',
            username: 'robica',
            email: 'robi@test.com',
        },{
            id: 3,
            name: 'Robo 2',
            username: 'robotto',
            email: 'robotto@test.com',
        }],
        searchField: 'Robo 1',
        isPending: false,
    };
    const wrapper2 = shallow(<App { ...mockProps2 } />);
    expect(wrapper2.instance().filterRobots()).toEqual([{
        id: 1,
        name: 'Robo 1',
        username: 'robica',
        email: 'robi@test.com',
    }]);
});

it('filters robots correctly 2', () => {
    const mockProps3 = {
        onRequestRobots: jest.fn(),
        robots: [{
            id: 1,
            name: 'Robo 1',
            username: 'robica',
            email: 'robi@test.com',
        },{
            id: 3,
            name: 'Robo 2',
            username: 'robotto',
            email: 'robotto@test.com',
        }],
        searchField: 'a',
        isPending: false,
    };
    const wrapper3 = shallow(<App { ...mockProps3 } />);
    expect(wrapper3.instance().filterRobots()).toEqual([]);
});
