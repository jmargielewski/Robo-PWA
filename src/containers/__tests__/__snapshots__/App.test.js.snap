// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render App without crashing 1`] = `
<div
  className="tc"
>
  <Header />
  <SearchBox />
  <Scroll>
    <ErrorBoundary>
      <CardList
        robots={Array []}
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
